fvm/env.go

```diff
@@ -244,22 +245,49 @@ func (e *hostEnv) CacheProgram(location ast.Location, program *ast.Program) erro
 	return e.ctx.ASTCache.SetProgram(location, program)
 }
 
-func (e *hostEnv) Log(message string) {
-	e.logs = append(e.logs, message)
+func (e *hostEnv) Log(message string) error {
+	if e.ctx.CadenceLoggingEnabled {
+		e.logs = append(e.logs, message)
+	}
+	return nil
 }
 
-func (e *hostEnv) EmitEvent(event cadence.Event) {
-	e.events = append(e.events, event)
-}
+func (e *hostEnv) EmitEvent(event cadence.Event) error {
 
-func (e *hostEnv) GenerateUUID() uint64 {
-	uuid, err := e.uuidGenerator.GenerateUUID()
+	payload, err := jsoncdc.Encode(event)
 	if err != nil {
-		// TODO - Return error once Cadence interface accommodates it
-		panic(fmt.Errorf("cannot get UUID: %w", err))
+		// TODO maybe we should return an error instead of panic
```

Comment:
I guess this todo is possible now since Emit can return an error

Reply:
applied at a49e36e